From f4defd1f46624cd3bca26ab677ea00e9a6ef405a Mon Sep 17 00:00:00 2001
From: Roopni Devanathan <rdevanat@qti.qualcomm.com>
Date: Mon, 8 Sep 2025 14:46:44 +0530
Subject: [PATCH] UPSTREAM:CP:iw: Set/get per-radio RTS threshold in
 multi-radio wiphy

In a multi-radio single-wiphy architecture, different radios can have
different RTS threshold requirements. Modify the iw command to get radio
index from user to operate on per-radio attributes in a multi-radio wiphy.
Modify the command such that the legacy userspace will still interact with
traditional drivers as well as multi-radio wiphy drivers. Also, print the
RTS threshold along with other per-radio attributes for each radio under
the section - "Supported wiphy radios" in iw phy#XXX info command.

In order to be able to set RTS threshold for a particular radio from user
space, without disturbing the other radios in a wiphy, pass the radio
index for which RTS threshold needs to be changed as optional pair of
arguments, along with its corresponding RTS threshold value. The valid
radio index values can be checked in iw phy#XXX info, under "Supported
wiphy radios". If radio index is not available, i.e., in case of single
radio wiphy, passing radio index is not required. If the radio index is not
provided, then the current behavior of setting passed RTS threshold to all
radios will be retained.

Command Usage:
iw phyX set rts <rts threshold|off> [radio <radio index>]

Sample output:
root@buildroot:~# iw phyXXX info
Wiphy phy0
        wiphy index: 0
        max # scan SSIDs: 16
        max scan IEs length: 339 bytes
        RTS threshold: 536
        Retry short limit: 7
        Retry long limit: 4
	.....
	Supported wiphy radios:
                * Idx 0:
                        RTS Threshold: 536
                        Frequency Range: 5170 MHz - 5835 MHz
			.....

                * Idx 1:
                        RTS Threshold: 231
                        Frequency Range: 2312 MHz - 2732 MHz
			.....

                * Idx 2:
                        RTS Threshold: 425
                        Frequency Range: 5945 MHz - 7125 MHz
			.....

	Globally valid interface combinations:
		.....

Patch-dependency: None
Patch-work: https://patchwork.kernel.org/project/linux-wireless/patch/20250718102659.111058-1-quic_rdevanat@quicinc.com/

Signed-off-by: Roopni Devanathan <rdevanat@qti.qualcomm.com>
---
 info.c |  3 +++
 phy.c  | 58 +++++++++++++++++++++++++++-------------------------------
 2 files changed, 30 insertions(+), 31 deletions(-)

diff --git a/info.c b/info.c
index c9c2b73..8621bd0 100644
--- a/info.c
+++ b/info.c
@@ -1102,6 +1102,9 @@ broken_combination:
 				case NL80211_WIPHY_RADIO_ATTR_INDEX:
 					printf("\t\t* Idx %u:\n", nla_get_u32(radio_prop));
 					break;
+				case NL80211_WIPHY_RADIO_ATTR_RTS_THRESHOLD:
+					printf("\t\t\tRTS Threshold: %d\n", nla_get_u32(radio_prop));
+					break;
 				case NL80211_WIPHY_RADIO_ATTR_FREQ_RANGE:
 					printf("\t\t\tFrequency Range: ");
 
diff --git a/phy.c b/phy.c
index f53cf56..7c48e16 100644
--- a/phy.c
+++ b/phy.c
@@ -617,56 +617,52 @@ static int handle_rts(struct nl80211_state *state,
 		      int argc, char **argv,
 		      enum id_input id)
 {
-	unsigned int rts, link_id;
-	int index = 0;
-	char *endptr;
+	unsigned int rts;
+	char *end;
 
-	if (!argc) {
+	if (argc != 1 && argc != 3)
 		return 1;
-	} else if (argc == 3) {
-		if (!strcmp(argv[0], "-l")) {
-			link_id = strtol(argv[1], &endptr, 10);
 
-			if (*endptr)
-				return 1;
-
-			if (link_id <= MAX_MLD_LINK) {
-				NLA_PUT_U8(msg, NL80211_ATTR_MLO_LINK_ID,
-					   link_id);
-				index = 2;
-			}
-		} else {
-			printf("Invalid parameter: %s\n", argv[0]);
+	if (strcmp("off", argv[0]) == 0)
+		rts = -1;
+	else {
+		if (!*argv[0])
+			return 1;
+		rts = strtoul(argv[0], &end, 10);
+		if (*end != '\0')
 			return 1;
-		}
 	}
 
-	if (strcmp("off", argv[index]) == 0) {
-		rts = -1;
-	} else {
-		char *end;
+	NLA_PUT_U32(msg, NL80211_ATTR_WIPHY_RTS_THRESHOLD, rts);
 
-		if (!*argv[index])
-			return 1;
+	argv++;
+	argc--;
 
-		rts = strtoul(argv[index], &end, 10);
+	if (argc > 1 && strcmp("radio", argv[0]) == 0) {
+		int radio_idx;
 
-		if (*end != '\0')
+		argv++;
+		argc--;
+
+		radio_idx = strtoul(argv[0], &end, 10);
+		if (*end != '\0')
 			return 1;
+
+		NLA_PUT_U8(msg, NL80211_ATTR_WIPHY_RADIO_INDEX, radio_idx);
+		argv++;
+		argc--;
 	}
 
-	NLA_PUT_U32(msg, NL80211_ATTR_WIPHY_RTS_THRESHOLD, rts);
+	if (argc)
+		return 1;
 
 	return 0;
  nla_put_failure:
 	return -ENOBUFS;
 }
-COMMAND(set, rts, "<rts threshold|off>",
+COMMAND(set, rts, "<rts threshold|off> [radio <radio_index>]",
 	NL80211_CMD_SET_WIPHY, 0, CIB_PHY, handle_rts,
 	"Set rts threshold.");
-COMMAND(set, rts, "[-l] <link_id> <rts threshold|off>",
-	NL80211_CMD_SET_WIPHY, 0, CIB_NETDEV, handle_rts,
-	"Set rts threshold.");
 
 static int handle_retry(struct nl80211_state *state,
 			struct nl_msg *msg,
-- 
2.34.1

